File: leetcode/cpp/p39-combination-sum.cpp
// LeetCode 30. Combination Sum.
#include <algorithm>
#include <iostream>
#include <vector>

#include "leetcode.hpp"

class Solution {
    std::vector<std::vector<int>> _combinations;
    int _target;
    void recurse(const std::vector<int>& candidates, size_t index, int tot, std::vector<int>& partials)
    {
        if (tot == _target) {
            _combinations.push_back(partials);
            return;
        }
        if (index >= candidates.size()) {
            return;
        }
        const int remainder = _target - tot;
        const int n = candidates[index];
        if (remainder < n) {
            return;
        }
        const size_t n_count = static_cast<size_t>(remainder / n);
        partials.insert(partials.end(), n_count, n);
        int m = n * n_count;
        while (true) {
            recurse(candidates, index + 1, tot + m, partials);
            if (m == 0) {
                break;
            }
            partials.pop_back();
            m -= n;
        }
    }

public:
    std::vector<std::vector<int>>
    combinationSum(std::vector<int>& candidates, int target)
    {
        std::sort(candidates.begin(), candidates.end());
        candidates.erase(std::find_if(candidates.begin(), candidates.end(), [=](const int i) { return i > target; }), candidates.end());
        _combinations.clear();
        _target = target;
        std::vector<int> partials;
        recurse(candidates, 0, 0, partials);
        return _combinations;
    }
};

int main()
{
    struct TestCase {
        std::vector<int> candidates;
        int target;
        std::vector<std::vector<int>> exp;
    };

    const TestCase test_cases[] = {
        { { 2, 3, 6, 7, 8 }, 7, { { 2, 2, 3 }, { 7 } } },
        { { 2, 3, 5 }, 8, { { 2, 2, 2, 2 }, { 2, 3, 3 }, { 3, 5 } } },
        { { 2 }, 1, {} },
    };

    Solution solution;
    for (const auto& tc : test_cases) {
        auto candidates_copy = tc.candidates;
        const auto act = solution.combinationSum(candidates_copy, tc.target);
        if (tc.exp != act) {
            std::cerr << "Fail. candidates: " << leetcode::to_string(tc.candidates)
                      << ", target: " << tc.target
                      << ", exp: " << leetcode::to_string(tc.exp)
                      << ", act: " << leetcode::to_string(act)
                      << "\n";
        }
    }
    return 0;
}

File: leetcode/py/p39-combination-sum.py
# LeetCode 39. Combination Sum.

import bisect
import copy
from typing import List


class Solution:
    def recurse(self, index, tot, partials):
        if index >= len(self.candidates):
            return
        remainder = self.target - tot
        n = self.candidates[index]
        if remainder < n:
            return
        m = remainder // n
        if remainder % n == 0:
            self.combinations.append(partials + [n] * m)

        for x in range(m, -1, -1):
            self.recurse(index + 1, tot + x * n, partials + [n] * x)

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        self.candidates = candidates[0 : bisect.bisect(candidates, target)]
        self.target = target
        self.combinations = []
        self.recurse(0, 0, [])
        return self.combinations


if __name__ == "__main__":
    test_cases = (
        ([2, 3, 6, 7, 8], 7, [[2, 2, 3], [7]]),
        ([2, 3, 5], 8, [[2, 2, 2, 2], [2, 3, 3], [3, 5]]),
        ([2], 1, []),
    )
    solution = Solution()
    for candidates, target, exp in test_cases:
        candidates_copy = copy.deepcopy(candidates)
        act = solution.combinationSum(candidates_copy, target)
        if exp != act:
            print(f"candidates: {candidates}, target: {target}, exp: {exp}, act: {act}")
